Deserialization for pre-v4 blocks

dorianvp opened the pull request:
On top of #429.

- Update `BlockHeaderData::parse_from_slice` to conditionally skip fields not present in pre-v4 block headers.
- Implement `FullTransaction::parse_vX` for older transaction formats.

dorianvp committed:
refactor(chain): prepare pre-v4 parsing

dorianvp committed:
Merge branch 'dev' into feat/block-deser

dorianvp committed:
chore: v1 & v2 tx parsing

dorianvp committed:
chore: remove parsing functions from `FullTransaction`

dorianvp committed:
chore: parse_v2 & v3 transactions

dorianvp committed:
chore: remove unused function

dorianvp committed:
chore: add `parse_joinsplits` helper

dorianvp committed:
chore: small lint fixes

dorianvp committed:
chore: run clippy

dorianvp committed:
chore: remove unused enum

dorianvp committed:
chore: add comments on unwraps

dorianvp committed:
Merge remote-tracking branch 'origin/dev' into feat/block-deser

dorianvp committed:
fix: cargo docs

dorianvp committed:
Merge branch 'dev' into feat/block-deser

dorianvp committed:
Merge branch 'dev' into feat/block-deser

dorianvp committed:
Merge branch 'dev' into feat/block-deser

AloeareV commented on zaino-fetch/src/chain/transaction.rs:
```diff
@@ -146,7 +146,46 @@ fn parse_transparent(data: &[u8]) -> Result<(&[u8], Vec<TxIn>, Vec<TxOut>), Pars
+/// - joinSplitSig: byte\[64\]
+///
+/// TODO: Enable skipped fields (all of them currently).
+fn parse_joinsplits(data: &[u8], version: u32) -> Result<(&[u8], Vec<JoinSplit>), ParseError> {
```
iiuc, we're reading joinsplits...and then after a JoinSplit is a pubkey and sig, which are separate and we ignore them?

dorianvp replied:
Updated!

AloeareV reviewed (commented):
Everything here that I understand looks great, I am confused by what's happening in the joinsplit code!

dorianvp committed:
Merge branch 'dev' into feat/block-deser

dorianvp committed:
fix(`TransactionData`): remove unnecessary parse function

dorianvp committed:
Merge branch 'dev' into feat/block-deser

dorianvp committed:
Merge branch 'dev' into feat/block-deser

zancas committed:
Merge branch 'dev' into feat/block-deser

dorianvp commented:
Adding tests. Will mark as ready for review once those land.

dorianvp committed:
chore: add initial block deser test

dorianvp committed:
Merge branch 'dev' into feat/block-deser

dorianvp committed:
fix: build issues

dorianvp committed:
Merge branch 'dev' into feat/block-deser

dorianvp committed:
Merge branch 'dev' into feat/block-deser

dorianvp committed:
chore: custom activation heights

idky137 commented on zaino-fetch/src/chain/transaction.rs:
```diff
@@ -422,6 +422,10 @@ struct TransactionData {
+    // /// Unix-epoch UTC time or block height, encoded as in Bitcoin (TODO: what does this mean?).
+    // ///
+    // /// Size\[bytes\]: 4
+    // lock_time: Option<u32>,
```
Why has an extra lock_time been added above the current NLockTime? Unless I am mistaken there is only one lock_time filed here so trying to skip or read extra bytes would break deserialisation.

dorianvp replied:
You're right, my mistake. Removed!

idky137 reviewed (dismissed):
Looks really close, I left one comment about the add lock_time field and I am a bit confused about the tests which have (or will be?) added. Are there anybtests that have been added to test deserialising v1, v2 and v3 blocks?

dorianvp committed:
Merge branch 'dev' into feat/block-deser

dorianvp committed:
chore: remove redundant comment

dorianvp committed:
chore: switch to zebrad

dorianvp commented:
I'm trying to do a `quick_shield`, but zingolib seems to be generating a transaction with the wrong transaction version:

```bash
thread 'varying_block_and_transaction_deserialization' panicked at integration-tests/tests/test_vectors.rs:709:41:
called `Result::unwrap()` on an `Err` value: SendError(TransmissionError(TransmissionFailed("Send Error: status: Internal, message: \"unhandled rpc-specific zaino_fetch::jsonrpsee::response::SendTransactionError error: RPC Error (code: -25): failed to validate tx: WtxId(\\\"private\\\"), error: transaction did not pass consensus validation: transaction version 5 not supported by the network upgrade Canopy\", details: [], metadata: MetadataMap { headers: {\"content-type\": \"application/grpc\", \"date\": \"Tue, 05 Aug 2025 20:42:50 GMT\", \"content-length\": \"0\"} }")))
```

dorianvp committed:
chore(block_deser): attempt with `zcashd`

dorianvp committed:
test(transaction_deser): add testvectors

nachog00 committed:
moved transaction test vectors into their own module and added v1,2 and 3 tvs

nachog00 committed:
Updated import path for test_vectors module
- Changed import path from `zaino_testutils::test_vectors::get_test_vectors` to `zaino_testutils::test_vectors::transactions::get_test_vectors` in `test_vectors.rs`

This change improves the organization of the test vectors by nesting them under the `transactions` module.

nachog00 committed:
Add new dependency to Cargo.lock
- Added "zaino-testutils" as a dependency in Cargo.lock

nachog00 committed:
feat(transaction): add comprehensive parsing tests for all transaction versions

nachog00 committed:
refactor(tests): use anyhow for error handling in deserialization test

nachog00 committed:
chore: update Cargo.lock

nachog00 committed:
filtered tests to only pre-v4

nachog00 committed:
Fix JoinSplit proof size parsing for v2/v3 transactions

- Fixed critical parsing bug where v2/v3 transactions weren't consuming all data
- Root cause: JoinSplit::parse_from_slice used incorrect proof sizes
  - v2/v3 use BCTV14 proofs (296 bytes)
  - v4 uses Groth16 proofs (192 bytes)
  - Parser was using 192 bytes for all versions, missing 208 bytes for v2/v3
- Made proof size version-dependent in JoinSplit parsing
- Updated all callers to pass transaction version to JoinSplit::parse_from_slice
- Improved error messages in unit tests to show remaining bytes
- All transaction parsing tests now pass (v1, v2, v3, v4)

nachog00 committed:
Improve test naming and documentation

- Renamed transaction parsing tests to clarify they use test vectors
- Updated test function names from test_real_vX_transaction_parsing to test_vX_transaction_parsing_with_test_vectors
- Updated test comments and error messages to remove "real blockchain" references
- Added documentation to test vector files to resolve missing_docs warnings

nachog00 committed:
Merge branch 'dev' into feat/block-deser

nachog00 committed:
cargo clippy and fmt
- test_vectors.rs: Replace `.len() > 0` with `.is_empty()` for checking collections
- transaction.rs: Improved error formatting and handling
- transaction.rs: Reformatted long lines for better readability
- zaino-testutils/src/lib.rs: Use `unwrap_or_default()` for cleaner code

nachog00 committed:
Merge branch 'feat/block-deser' of github.com:zingolabs/zaino into feat/block-deser

nachog00 committed:
fixed clippy warnings
- Updated error message formatting in `JoinSplit::parse_from_slice`
- Enhanced readability by utilizing inline formatting in `skip_bytes` function

AloeareV committed:
move test vectors to new crate to avoid cyclic dependency in unit tests

AloeareV committed:
clippify

AloeareV reviewed (approved):
Fantastic!
